### ISSUE TITLE
Discussion: should __cuda_array_interface__ include a stream pointer?

### ISSUE THREAD

**Issue by leofang** [numba/numba#4933]
**Context:** TensorFlow is reluctant in supporting `__cuda_array_interface__` because they run everything on non-default streams (see discussion around https://github.com/tensorflow/tensorflow/issues/29039#issuecomment-527520270). 

**Question:** should `__cuda_array_interface__` include a pointer to the stream on which the data is being changed? 

**My opinion: No.** In that case we just do a stream synchronization on the producer side (i.e., using the producer's stream API) before passing the array to the consumer. In the worst scenario, we can do a device synchronization manually at the interoperating point. I don't see why this is a concern preventing them from supporting the protocol, so I want to hear from Numba folks if you have any thoughts. 

Thanks.

**Comment by VoVAllen** [comment]
Is `__cuda_array_interface__` an async API to hold the memory pointer?

**Comment by nouiz** [comment]
The producer frameworks can always sync as you said. So it is not strictly needed. I would wait to see the need in practice before making a new version of the protocol that will be more complicated and would need update from all frameworks that already adopted it.
There is tons of benefit from going without that protocol to implementing it. Going from the current protocol version to the stream version have only theoretical gain to my knowledge and I expect them to be low or none in practice. I also think people would easily understand that changing of frameworks have some cost. So it shouldn't surprise many users.

**Comment by seibert** [comment]
Our goal with `__cuda_array_interface__` was to follow NumPy's `__array_interface__` in providing a standard way to expose the pointer and the dtype.  It doesn't try to solve issues like synchronization or memory lifetime, which vary too much between frameworks, and we assumed would be handled by users who know their actual application needs.

A more sophisticated protocol for managing memory lifetime and synchronization would be useful, but I think outside the scope of what we the Numba project could propose.  Honestly, I'd like to see this as part of a full blown memory pool manager, like RMM.

**Comment by dhirschfeld** [comment]
That sound similar to the goals of `umem`:
https://github.com/xnd-project/umem

**Comment by leofang** [comment]
Thank you all for sharing your thoughts. I'd like to echo what @seibert said: 
> we assumed would be handled by users who know their actual application needs

I think this is a necessary assumption for general discussions of the CUDA array interface. 

From what I read I think we agreed it's an overkill to add a stream pointer, so I am closing this issue. Please feel free to reopen if you want to keep the discussion alive and/or wanna share different views. Thanks to all again!

**Comment by sanjoy** [comment]
I'm coming here from the [TF discussion of `__cuda_array_interface__`](https://github.com/tensorflow/tensorflow/issues/29039).

I think there are two orthogonal issues here.

The first is one is about having an unambiguous specification -- what does framework $FOO need to do around streams to get a minimally correct implementation?  It seems like in practice the contract is that the memory needs to be available on the default stream during conversion to and from via `__cuda_array_interface__`.  If this is correct, I'd suggest we codify just that.  TF would have to conform to this interface like all the other frameworks listed here.

The second issue is about what needs to happen to provide optimal performance for TensorFlow.  We can revisit this later once we have something working and have evidence of what actually matters.

**Comment by leofang** [comment]
Since @sanjoy joined as part of the TF team, I am reopening the issue for the time being.

**Comment by sanjoy** [comment]
Continuing from the [discussion](https://github.com/tensorflow/tensorflow/issues/29039#issuecomment-569294745) on the TF issue:

> I incline to simply say the user is responsible to ensure the memory is ready at the interoperating point. 

We can do this if we have to, but it doesn't sound particularly user friendly.  This requires the user to understand the internal implementation details for the two interacting libraries to pass tensors from one library to the other.  Moreover, these implementation details may change over time which is another problem.  For instance, a framework that only uses a default stream today might want to use multi-stream execution in the future.

**Comment by sklam** [comment]
I agree with @sanjoy that we can't assume the user to know that much about the implementation details. In addition, the concept of the default stream depends on what nvcc compilation flag is used, the legacy default stream is the NULL stream, which is implicitly synchronizing for all host threads. With the `--default-stream per-thread` flag, the default stream is different per thread and it is no longer implicitly synchronizing. I also worry that nvcc may change the default stream soon that the old behavior is now called *legacy*. (ref: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#default-stream)

I'd prefer that, at the time of export, the producer must have synchronized all operations on the exported memory. In other words, exporting the `__cuda_array_interface__` is an synchronizing operation. That way, it doesn't matter whether the consumer is reading on the same stream or from the same host thread.

**Comment by sanjoy** [comment]
> I'd prefer that, at the time of export, the producer must have synchronized all operations on the exported memory. In other words, exporting the `__cuda_array_interface__` is an synchronizing operation.

That's fine by me as long as it is explicitly documented and all frameworks implementing `__cuda_array_interface__` follow this invariant.

**Comment by leofang** [comment]
Not sure if you guys are waiting for further comments. I agree that this needs to be explicitly documented. But instead of 

> I'd prefer that, at the time of export, the producer must have synchronized all operations on the exported memory.

I would say "the user" must do the synchronization. Not the producer, not the consumer. This is in part for performance reasons, but also because, like @seibert commented above:

> we assumed would be handled by users who know their actual application needs

**Comment by sanjoy** [comment]
I'm fine with documenting that it is up to the user to insert the necessary synchronization.

Personally, I think this is a tricky invariant for a framework *user* to maintain (for reasons I mentioned [above](https://github.com/numba/numba/issues/4933#issuecomment-569303059)) but ultimately this is something that's up to the the authors for `__cuda_array_interface__` to decide.

**Comment by nouiz** [comment]
I agree with @sanjoy argument, but not the conclusion. I consider it a mistake if it is error-prone by default. Do not forget, end users are data scientists, not computer scientist. They do not want to care about that level of detail.

Another point to @sanjoy argumentation. If the user test it his code, it could work 99% of the time, but still fail 1% of the time. Random failures are hard to debug for developers, so even harder for the end user.

It should be 100% correct by default. Maybe an option to enable an unsafe mode for very advanced users could be made. But at that point, I think it would be better to have a new version of the interface where the producer frameworks that expose the stream would be better. This way, the consumer frameworks would always do the best and it would always be correct.

Do not forget that the speed up of the 100% correct version vs the current state (the copy to/from the CPU) is much bigger than the speed up from removing an extra sync. If that extra sync saving is needed, a 100% correct way to get it should be designed.

**Comment by nouiz** [comment]
This interface was designed by the Numba team.
@sklam is the last person working on Numba that commented on that thread. And my understanding of his comment is that the implementation must be safe by default. (So sync what is needed to guaranty the safe usage).
After this comment, @leofang (not part of the Numba team to my knowledge) made another comment that point in the other direction.

As this is an important design decision and to prevent misinterpretation, can we have a confirmation of what is the official decision on this?

**Comment by leofang** [comment]
I guess the Numba team is gonna freeze code in a few days for 0.48rc. Please don’t rush into any decision at this point. Last time we revised the v2 protocol a few hours before the cutoff time, and encountered a lot of setbacks and one bug report afterwards. 😥

**Comment by stuartarchibald** [comment]
I'll raise this at the Numba core developer meeting on Tuesday (14 Jan 2020) unless resolved before hand.

**Comment by leofang** [comment]
@jakirkham @gmarkall Would this discussion be relevant to the NBEP 7 proposal that you're drafting?

**Comment by sklam** [comment]
After some discussion at the numba dev meeting, @gmarkall suggested that we should understand the use-cases more.  For instance, when using with TF being the producer, would the sync become a major performance limiter and defeating the purpose of using Numba?

**Comment by gmarkall** [comment]
@leofang Good question - in order to try not to tackle too many related issues at once, I was hoping to avoid suggesting changes to `__cuda_array_interface__` as part of it, and only focus on using it to change Numba's internal allocation. There was an issue raised that the `stream` parameter might need more consideration if there's a future asynchronous allocator: https://github.com/gmarkall/nbep-7/issues/7 - i.e. it would be important not to operate on an allocation from a different stream without a full synchronization first, lest the allocation hasn't completed before the operation on it. However, as that functionality doesn't presently exist, I'm hoping to avoid actively considering it as part of NBEP 7 and would prefer to leave it for future versions of the EMM Plugin interface.

I think the discussion is relevant in the long term - but for now I hope it will keep simpler for both discussions to keep them separate.

**Comment by gmarkall** [comment]
Having thought over this a bit, I'm inclined to agree with @sanjoy's earlier points:

> The first is one is about having an unambiguous specification -- what does framework $FOO need to do around streams to get a minimally correct implementation? It seems like in practice the contract is that the memory needs to be available on the default stream during conversion to and from via `__cuda_array_interface__`. If this is correct, I'd suggest we codify just that. TF would have to conform to this interface like all the other frameworks listed here.
> 
> The second issue is about what needs to happen to provide optimal performance for TensorFlow. We can revisit this later once we have something working and have evidence of what actually matters.

Since it sounds like there aren't any concretely-defined use cases as yet because TF has not yet adopted `__cuda_array_interface__` , it seems that the straightforward path to progressing on this issue, and to getting a better understanding of actual use cases, is:

- Immediately codifying that memory needs to be available on the default stream during conversion to and from via `__cuda_array_interface__` so that TF can implement it.
- Revisiting this discussion in future once we have some performance analysis showing bottlenecks related to synchronization when importing / exporting via `__cuda_array_interface__` that would be resolved by adding the stream pointer.

(I realise I'm mostly just re-stating @sanjoy's points there :-) )

I think this provides a safe specification for now, in the sense that it would be hard to mis-use by either the framework developer or user - so I think this also addresses @nouiz's concerns above.

From @leofang's comment above, though:

> I would say "the user" must do the synchronization. Not the producer, not the consumer. This is in part for performance reasons, but also because, like @seibert commented above:
>> we assumed would be handled by users who know their actual application needs

> By doing so, I think we won't need to touch the existing implementations of __cuda_array_interface__ in various libraries, and only need to add a warning in the protocol documentation.

I understand the above plan would mean that existing implementations would need to be changed to ensure the memory is available on the default stream at the time of export. Could this be addressed by also bumping the `__cuda_array_interface__` version number along with codifying that the memory needs to be available on the default stream? I think at least this wouldn't mean mismatched expectations between existing implementations.

To the more general points about the goals of the interface and expectations of the user brought up by @seibert and @leofang: Unfortunately I haven't been present in the discussions around the `__cuda_array_interface__` and its implementation until recently, so there will be historical discussions and details that I'm not aware of. I understand that the original goal was to provide only a standard way to expose the pointer and dtype and to consider synchronization issues out of scope. It seems like the original goal has been (at least mostly achieved), and is now in a way starting to impose some limitations on interoperability. I also don't see a straightforward way to easily push the issue into an external memory manager without a lot of coordination between framework authors. Could it be acceptable to re-consider the goals of the array interface to include synchronization concerns?

**Comment by leofang** [comment]
Thank you, Graham, for revisiting this thread with more thoughts. I pinned you also for the same reason: 
> I also don't see a straightforward way to easily push the issue into an external memory manager without a lot of coordination between framework authors.

To make NBEP 7 more acceptable, it'd be better to consider the async behavior by including streams. (My impractical wish was to also solve the issue #4797 in along this discussion, but it'd be too much work...)

Anyway, back to the main topic. The concern is understood, but from my point of view the synchronization just hurts the performance too much. For example, we have folks running CuPy + Numba + MPI routinely on Summit (the largest supercomputer in the world). Try synchronizing a billion or more times every time a Numba kernel is called. Those people will hate us 🙁 

If bumping the protocol version is inevitable, I wonder if the following addition would be appreciated: we optionally include a stream pointer along with a flag indicating whether a synchronization over the stream should be performed by the consumer or not (and it defaults to yes for safety). If it's `None`, then it means the memory is exported on the (legacy) default stream. A draft of this proposed change is given below: 
```diff
diff --git a/docs/source/cuda/cuda_array_interface.rst b/docs/source/cuda/cuda_array_interface.rst
index aee84dd..6d4f328 100644
--- a/docs/source/cuda/cuda_array_interface.rst
+++ b/docs/source/cuda/cuda_array_interface.rst
@@ -48,7 +48,7 @@ that must contain the following entries:
 - **version**: `integer`
 
   An integer for the version of the interface being exported.
-  The current version is *2*.
+  The current version is *3*.
 
 
 The following are optional entries:
@@ -76,6 +76,16 @@ The following are optional entries:
             and will raise a `NotImplementedError` exception if one is passed
             to a GPU function.
 
+- **stream**: ``None`` or `(integer, boolean)`
+
+  If **stream** is not given, or it is ``None``, the array is exported on
+  the (legacy) default stream. Otherwise, a 2-tuple is given: the first
+  element is the pointer to the stream (as a Python `int` or `long`) on
+  which the array is exported, and the second element is a `bool` flag
+  indicating whether the consumer is expected to synchronize over the
+  stream or not (the consumer can choose not to, see below), which
+  defaults to `True`.
+
 Additional information about the data pointer can be retrieved using
 ``cuPointerGetAttribute`` or ``cudaPointerGetAttributes``.  Such information
 include:
```
The decision for whether to sync should be made by the consumer for obvious reason: Only the consumer knows if there's a performance or correctness concern or not.

To make use of the flag, like in an HPC setup, we could have a named variable added to every supporting library, something like `cuda_array_interface_stream_sync`, so that the flag can be ignored at runtime in the following way:
```python
import some_lib
some_lib.cuda_array_interface_stream_sync = False  # turn off sync 

# in some_lib's import function:
# ...do some setup and initialization
ptr = input_arr.__cuda_array_interface__['data'][0]
stream_ptr, sync_flag = input_arr.__cuda_array_interface__['stream']
if some_lib.cuda_array_interface_stream_sync and sync_flag:
    some_lib.cuda.stream_sync(stream_ptr)
# ...do other work to wrap the raw memory and return
```
What do you guys think?

**Comment by gmarkall** [comment]
@leofang many thanks for adding the details of the use case:

> The concern is understood, but from my point of view the synchronization just hurts the performance too much. For example, we have folks running CuPy + Numba + MPI routinely on Summit (the largest supercomputer in the world). Try synchronizing a billion or more times every time a Numba kernel is called. Those people will hate us 🙁

I wasn't aware that there were use cases like this in practical use already (sorry if I had missed an explanation of this before). I can see why just codifying that a synchronization is required make an immediate practical problem for this use of the interface.

I'll think a bit more about your suggested proposal and comment back later.

> To make NBEP 7 more acceptable, it'd be better to consider the async behavior by including streams. (My impractical wish was to also solve the issue #4797 along this discussion, but it'd be too much work and trouble...)

I've created an issue in the NBEP-7 repo to further this discussion a bit without adding too much NBEP-7 noise in this issue: https://github.com/gmarkall/nbep-7/issues/14

**Comment by sklam** [comment]
@gmarkall, i am wondering if we can provide a safe behavior for the situation where the producer is using streams with the current interface (lack of stream pointers). What if we say that `__cuda_array_interface__` consumer are expected to operate on the legacy default stream for its implicit synchronization. If we can lock down the spec for a safe behavior now, we can unblock other libraries from implementing the interface. The stream pointer question can then be a optimization.

**Comment by leofang** [comment]
> I wasn't aware that there were use cases like this in practical use already (sorry if I had missed an explanation of this before). I can see why just codifying that a synchronization is required make an immediate practical problem for this use of the interface.

No worries, @gmarkall 🙂 I don't think I brought up this use case in this thread. 

Thanks @sklam for reminding me the legacy stream. I updated my pseudo code at the bottom of my proposal (https://github.com/numba/numba/issues/4933#issuecomment-578088647). I meant to do that (as can be seen in the proposed diff) but forgot due to lack of sleep...😫 

One drawback of my proposal that I can think of is that if the consumer library does not provide a wrapper for CUDA functionalities (like in mpi4py), my pseudo code won't work, as there is no function `some_lib.cuda.stream_sync()` for us to call in [the internal import function](https://github.com/mpi4py/mpi4py/blob/869edd72a26a6406ba0d967a49c645773cf4bad0/src/mpi4py/MPI/asgpubuf.pxi#L59-L129), and we will have to rely on the default behavior.

> What if we say that `__cuda_array_interface__` consumer are expected to operate on the legacy default stream for its implicit synchronization. 

This would work as a short-term solution. In practice it is not against [what I've suggested earlier](https://github.com/numba/numba/issues/4933#issuecomment-570797858).

> If we can lock down the spec for a safe behavior now, we can unblock other libraries from implementing the interface.

You mean for TF, not for other libraries that already support the protocol, right?

**Comment by gmarkall** [comment]
> @gmarkall, i am wondering if we can provide a safe behavior for the situation where the producer is using streams with the current interface (lack of stream pointers). What if we say that `__cuda_array_interface__` consumer are expected to operate on the legacy default stream for its implicit synchronization. If we can lock down the spec for a safe behavior now, we can unblock other libraries from implementing the interface. The stream pointer question can then be a optimization.

This does sound like it is aligned with @sanjoy's description of how this could work in TF on the TF issue: https://github.com/tensorflow/tensorflow/issues/29039#issuecomment-569182709

**Comment by gmarkall** [comment]
In #5162 I've attempted to revise the specification of the array interface to include the semantics @sklam suggested (though on the default stream rather than the legacy default stream). One extra thing I felt I had to consider was what the producer must to do ensure that the consumer's behaviour of operating on the default stream safe, which is also discussed in the specification changes in that PR.

**Comment by leofang** [comment]
I am closing this as this is now specified in the v3 protocol (#5162). Feel free to reopen if there's anything we left unaddressed. Thanks again to everyone and especially @gmarkall for all the hard work!
